### PR TITLE
[OptApp] Adding SciPy Library

### PR DESCRIPTION
**Pull request by Igarizza**
**📝 Description**
PR adds possebility to use SciPy as optimizer.

### CODE REVIEW

File: applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+from KratosMultiphysics.OptimizationApplication.utilities.logger_utilities import time_decorator
+import scipy.optimize
+import numpy as np
+from scipy.optimize import NonlinearConstraint
+from KratosMultiphysics.OptimizationApplication.utilities.optimization_problem_utilities import OutputGradientFields
+
+try:
+    import scipy
+except ImportError:
+    raise Exception("SciPy python library is not available")
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py]: You need to move the above two scipy imports to the same try catch block.

File: applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            "objective"         : {},
+            "constraints"         : [],
+            "controls"          : [],
+            "echo_level"        : 0,
+            "SciPy_settings"          : {
+                "method"      : "",
+                "lower_bound" : -1e16,
+                "upper_bound" : 1e16,
+                "options"     : {}
+                }
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py]:
```suggestion
            "module"            : "KratosMultiphysics.OptimizationApplication.algorithms",
            "type"              : "PLEASE_PROVIDE_AN_ALGORITHM_CLASS_NAME",
            "objective"         : {},
            "constraints"       : [],
            "controls"          : [],
            "echo_level"        : 0,
            "SciPy_settings"    : {
                "method"      : "",
                "lower_bound" : -1e16,
                "upper_bound" : 1e16,
                "options"     : {}
            }
```

File: applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            res = scipy.optimize.minimize(self.__objective.CalculateStandardizedValue, self.x0, method=self.SciPy_settings["method"].GetString(), jac=self.__objective.CalculateStandardizedGradient, constraints=self.__scipy_constraints, options=self.__GetOptions(), callback=self.Output, bounds=self.bounds)
+        print("res::success: ", res.success)
+        print("res::status: ", res.status)
+        print("res::message: ", res.message)
+
+    @time_decorator()
+    def Output(self, *args) -> KratosOA.CollectiveExpression:
+        if self.__objective.computed:
+
+            print(f"Output iteration {self._optimization_problem.GetStep()}")
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py]:
```suggestion
            Kratos.Logger.PrintInfo(self.__class__.__name__, f"Output iteration {self._optimization_problem.GetStep()}")
```

File: applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+    @time_decorator()
+    def Solve(self):
+        if not self.__scipy_constraints:
+            res = scipy.optimize.minimize(self.__objective.CalculateStandardizedValue, self.x0, method=self.SciPy_settings["method"].GetString(), jac=self.__objective.CalculateStandardizedGradient, options=self.__GetOptions(), callback=self.Output, bounds=self.bounds)
+        elif self.__scipy_constraints:
+            res = scipy.optimize.minimize(self.__objective.CalculateStandardizedValue, self.x0, method=self.SciPy_settings["method"].GetString(), jac=self.__objective.CalculateStandardizedGradient, constraints=self.__scipy_constraints, options=self.__GetOptions(), callback=self.Output, bounds=self.bounds)
+        print("res::success: ", res.success)
+        print("res::status: ", res.status)
+        print("res::message: ", res.message)
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py]:
```suggestion
        Kratos.Logger.PrintInfo(self.__class__.__name__, f"res::success: {res.success}")
        Kratos.Logger.PrintInfo(self.__class__.__name__, f"res::status: {res.status}")
        Kratos.Logger.PrintInfo(self.__class__.__name__, f"res::message: {res.message}")
```

File: applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            for process in self._optimization_problem.GetListOfProcesses("output_processes"):
+                if process.IsOutputStep():
+                    process.PrintOutput()
+            
+            # Advance in Optimization Iteration
+            self._optimization_problem.AdvanceStep()
+            self.__objective.computed = False
+
+    def __GetOptions(self):
+        Kratos.Parameters
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py]:
```suggestion
```

File: applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        dict_options = dict()
+        for key in options.keys():
+            if options[key].IsBool():
+                dict_options[key] = options[key].GetBool()
+            elif options[key].IsDouble():
+                dict_options[key] = options[key].GetDouble()
+            elif options[key].IsInt():
+                dict_options[key] = options[key].GetInt()
+            elif options[key].IsString():
+                dict_options[key] = options[key].GetString()
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py:156]:
```suggestion
            elif options[key].IsString():
                dict_options[key] = options[key].GetString()
            else:
                raise RuntimeException(f"Unsupported data type with key = \"{key}\" [ value = {options[key]} ].")
```

File: applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    @classmethod
+    def GetDefaultParameters(cls):
+        return Kratos.Parameters("""{
+            "module"            : "KratosMultiphysics.OptimizationApplication.algorithms",
+            "type"              : "PLEASE_PROVIDE_AN_ALGORITHM_CLASS_NAME",
+            "objective"         : {},
+            "constraints"         : [],
+            "controls"          : [],
+            "echo_level"        : 0,
+            "SciPy_settings"          : {
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py]: Not a big fan of using mixed camel and snake case. But I understand, `sci_py_settings` does not make senses ryt?

**Review by sunethwarna — COMMENTED** [review]
Thanks @igarizza for this nice addition. :). I only have minor comments.

File: applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py (hunk 2 above)
Comment (Igarizza) [reply applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py]: That would be a bit ugly in the code, I will explain in person.

File: applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py (hunk 7 above)
Comment (Igarizza) [reply applications/OptimizationApplication/python_scripts/algorithms/SciPy_algorithms.py]: they text their name as SciPy and it looks nice=) that is the only concern I have. Can I keep it or you suggest change?

**Review by talhah-ansari — APPROVED** [review]
Nice. Thanks @Igarizza

**Pull request merged**